### PR TITLE
`Development`: Refactor quiz exercise submission

### PR DESCRIPTION
**Pull request by rriyaldhi**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Most of the implementation of Quiz Submission (used by existing Quiz Exercise) will be reused for Quiz Exam Submission (used for Quiz Questions from Quiz Pool that will be added in a separate PR). To avoid code duplication, some of the implementations of Quiz Submission need to be abstracted.

### Description
1. Created AbstractQuizSubmission model class and moved common properties of QuizSubmission and QuizExamSubmission to this model class
2. Modified SubmittedAnswer model to depend on AbstractQuizSubmission instead of QuizSubmission
3. Created AbstractQuizSubmissionService and moved the common codes of QuizSubmissionService (used by existing Quiz Exercise) and QuizExamSubmissionService (Quiz Questions from Quiz Pool that will be added in a separate PR) to this class
4. Refactor the calculateAndUpdateScores method to receive a  list of QuizQuestion instead of QuizExercise, so that the method can be reused by QuizExamSubmissionService later.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Quiz Exercise (Can be any type: Synchronize, Batched, or Individual)

1. Log in to Artemis as a Student
5. Participate in the Quiz Exercise
6. Verify that the answers are saved by refreshing the page after submitting and ensure that the answers are saved
7. If Quiz is Batched or Individual, click **End Now** as Instructor. Or wait until the time ends for a Synchronized Quiz
8. Verify that the students can see the result
9. As an instructor, verify that the preview mode of the Quiz Exercise is still working by clicking the Preview Quiz button, submitting the answers, and verifying that the results are shown
10. As an instructor, verify that the solution mode of the Quiz Exercise is still working by clicking the Sample Solution button and verifying that the solutions are displayed

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a Quiz Exercise

1. Log in to Artemis as a Student
2. Participate in the exam
3. Click the **Save** button
4. Refresh the page and verify that the answers are saved
5. Submit the exam
11. Log out and log in as an Instructor
12. Navigate to the Exam and click Assessment
13. Wait until the exam is finished and the grace period has ended, then click Evaluate Quiz Exercises
14. Log out and log in as a Student
15. Verify that the students can see the result

#### Code Review
- [x] Review 1
- [ ] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| AbstractQuizSubmission.java | 80% | 92% |
| QuizSubmission.java | n/a | 100% |
| SubmittedAnswer.java | 50% | 91% |
| AbstractQuizSubmissionService.java | n/a | 88% |
| ExerciseService.java | 81% | 96% |
| QuizExerciseService.java | 85% | 96% |
| QuizSubmissionService.java | 73% | 83% |
| ExamQuizService.java | 76% | 87% |
| QuizScheduleService.java | 72% | 89% |
| ParticipationResource.java | 69% | 91% |
| QuizSubmissionResource.java | 93% | 100% |
| quiz-exam-submission.component.ts | 77.61% | 92.72% |

### Screenshots

**Exercise**
https://user-images.githubusercontent.com/10885503/231222471-1037787a-5ab8-488c-904e-48aaa6d76957.mov

**Exam**
https://user-images.githubusercontent.com/10885503/231222458-affbc09f-7198-49fa-b13f-8de0098bf03c.mov

### CODE REVIEW

**Review by TheZoker — DISMISSED** [review]
Tested within testing session on TS2, LGTM 👍

**Review by tobias-lippert — DISMISSED** [review]
Tested in testing session on ts2 with a course quiz exercise and an exam quiz exercises
In both cases everything worked as described.

File: src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java:201]: Maybe rename to `hasUserSubmitted`. I dont know if login as a String parameter makes sense. Is there no way to use user instead?

**Review by TimOrtel — DISMISSED** [review]
Thank you for this PR. I checked the code which seems good. 

It looks like you changed the API behavior in this PR, therefore this also affects the Android version. Can you please either create a PR in the Android repository that applies these changes to the Quiz feature or create an issue and link this PR so we do not forget to implement this once merged?

**Comment by rriyaldhi** [comment]
> Thank you for this PR. I checked the code which seems good. 
> 
> It looks like you changed the API behavior in this PR, therefore this also affects the Android version. Can you please either create a PR in the Android repository that applies these changes to the Quiz feature or create an issue and link this PR so we do not forget to implement this once merged? 
> 
> 

Thanks for noticing. Sure, I'll create an issue first so that we won't forget about this.

File: src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java]: yes, hasUserSubmitted is more clear. But the login parameter is required as the information is expected to be defined by the caller.

**Review by frederik-kt — DISMISSED** [review]
Tested on TS. Works as expected.

**Review by DominikRemo — DISMISSED** [review]
tested in testing session on ts 2. Works as expected

**Review by 1999Markus — DISMISSED** [review]
Tested in testing session!

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/AbstractQuizSubmission.java
Comment (frederik-kt) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/AbstractQuizSubmission.java:74]:
```suggestion
     * It sets the {@link #scoreInPoints} of the submission to null.
     * Additionally, it calls {@link SubmittedAnswer#filterOutCorrectAnswers()} for every submitted answer.
```

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/AbstractQuizSubmission.java
Comment (frederik-kt) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/AbstractQuizSubmission.java:105]:
```suggestion
     * @param quizQuestions the quiz questions of the {@link QuizExercise} this submission belongs to. These questions are needed to have values for isCorrect in the answer options.
```
Also maybe explain in a little more detail why the questions are needed. I am not sure how `isCorrect` ties in with the method implementation.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/AbstractQuizSubmission.java
Comment (frederik-kt) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/AbstractQuizSubmission.java:126]: There's some duplication between this and . You should be able to just call `AbstractQuizSubmission#getScoreInPoints()` from `QuizExercise#getScoreInPointsForSubmission()` right?

File: src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java
Comment (frederik-kt) [inline src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java:228]: Maybe add a comment why the participation is set here? Calling this method it is not clear why you would have to provide the quizExercise and the user to save the quizSubmission.

File: src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java
Comment (frederik-kt) [inline src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java:237]: Maybe add a comment why the participation is set here? Calling this method it is not clear why you would have to provide the quizExercise and the user to save the quizSubmission.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/AbstractQuizSubmission.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/AbstractQuizSubmission.java]: Updated the JavaDoc

File: src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java]: Yes, it happens in AbstractQuizSubmissionService#saveSubmissionForExamMode(). Updated the JavaDoc

File: src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java:237]: Yes, agree. The JavaDoc is updated.

**Review by TheZoker — APPROVED** [review]

**Review by frederik-kt — DISMISSED** [review]
Code looks good. Also tested on TS2 and everything worked as expected.

**Review by greekgeek1 — DISMISSED** [review]

**Review by TheZoker — APPROVED** [review]

**Review by frederik-kt — APPROVED** [review]
Re-approval after merge conflict. Also tested on TS2 and still works as expected.

**Review by MaximilianAnzinger — APPROVED** [review]
Tested on TS2 and works as expected.

**Review by YeetTheFirst21 — DISMISSED** [review]
Tested on TS2, all seems to work as expected

**Review by Santia-go — DISMISSED** [review]
Tested on TS1

All work as expected. 
- The creation of a quiz with different types of questions works
- Participating as the student works also 
- The live statistics works also 
- I just found a little bug that is related
- The questions are saved after submission and when the time is up.  
- The preview view for the instructor works 
- The scores are correct from both sides (student and instructor)
- The sample solution is also working for all questions and quizzes. 

The same occurs in the exam mode. 
- Creation and participation is working properly. 
- Grades and results are saved

**Review by frederik-kt — APPROVED** [review]
Tested in testing session on TS4. Everything still works as expected.

**Review by pellumbbaboci — APPROVED** [review]
Tested during testing session on test4 works as expected

**Review by YeetTheFirst21 — APPROVED** [review]
Approved again after the changes

**Review by Santia-go — APPROVED** [review]
Re-approved after solving the merge conflict.

**Review by DominikRemo — APPROVED** [review]
Re-approved

**Pull request merged**
